torchmultimodal/models/gpt.py

```diff
@@ -28,12 +29,278 @@ class TransformerLayerOutput(NamedTuple):
+            head_mask=head_mask,
+            use_cache=use_cache,
+            causal=causal,
+            right_shift=right_shift,
+            return_attn_weights=return_attn_weights,
+            return_hidden_states=return_hidden_states,
+        )
+
+        hidden_states = decoder_output.last_hidden_states
+        logits = self.head(hidden_states, logits_mask)
+
+        return MultimodalGPTOutput(decoder_output, logits)
+
+    def fwd(
+        self,
+        in_tokens: Optional[Tensor] = None,
+        out_tokens: Optional[Tensor] = None,
+        in_pos_ids: Optional[Tensor] = None,
+        out_pos_ids: Optional[Tensor] = None,
+        attn_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        use_cache: bool = False,
+        causal: bool = False,
+        right_shift: bool = False,
+        return_attn_weights: bool = False,
+        return_hidden_states: bool = False,
+    ) -> TransformerDecoderOutput:
+        # During training this method is used in the forward pass to decode input- and
+        # output- tokens.
+        # During generation this method is used for autoregressive decoding.
+        if (in_tokens is None) and (out_tokens is None):
+            raise ValueError(
+                "input-modality token and output-modality token sequences cannot be both empty"
+            )
+
+        # Look up embeddings for the given tokens and project to fit the size of the
+        # transformer decoder
+        in_modality = out_modality = None
+
+        if in_tokens is not None:
+            # (b, in_seq_len, in_emb_dim)
+            in_modality = self.lookup(in_tokens, "in")
+            if self.in_projection is not None:
+                in_modality = self.in_projection(
+                    in_modality
+                )  # (b, in_seq_len, d_model)
+        if out_tokens is not None:
+            # (b, out_seq_len, out_emb_dim)
+            out_modality = self.lookup(out_tokens, "out")
+            if self.out_projection is not None:
+                out_modality = self.out_projection(
+                    out_modality
+                )  # (b, out_seq_len, d_model)
+
+        return self.mm_decoder(
+            in_modality=in_modality,
+            out_modality=out_modality,
+            in_pos_ids=in_pos_ids,
+            out_pos_ids=out_pos_ids,
+            attn_mask=attn_mask,
+            head_mask=head_mask,
+            use_cache=use_cache,
+            causal=causal,
+            right_shift=right_shift,
+            return_attn_weights=return_attn_weights,
+            return_hidden_states=return_hidden_states,
+        )
+
+    def head(
+        self, hidden_states: Tensor, logits_mask: Optional[Tensor] = None
+    ) -> Tensor:
+        out = self.norm(hidden_states)
+        logits = self.to_logit(hidden_states)
+        max_neg_value = -torch.finfo(logits.dtype).max
+        if logits_mask is not None:
+            logits.masked_fill_(logits_mask == 0, max_neg_value)
```

Comment:
Maybe avoid the inplace version (unless you have a good reason for it). [Ref](https://lernapparat.de/pytorch-inplace)